testbook/client.py

```diff
@@ -7,12 +7,28 @@
 
 from .exceptions import TestbookCellTagNotFoundError, TestbookError
 from .testbooknode import TestbookNode
+from .translators import PythonTranslator
+from .reference import TestbookCallableReference, TestbookVariableReference
+from .utils import random_varname, _construct_call_code
+
+_testbook_object_cache = dict()
```

Comment:
I've implemented a caching system to store references, so that new reference objects don't get created for the same kernel variable/function.

Reply:
> Careful here, you just created a global cache object that's not scoped to a particular notebook. If you're going to cache make it an object on the TestbookNotebookClient instance.

Right! I didn't think of that! I will move it to the instance instead.

> Also if those references have state themselves this is doubly dangerous and you may not want to cache at all. 

I think this shouldn't be a problem because the reference object just holds a reference to the variable name in the kernel and not the value itself.. 

> What behavior was leading you to wanting a cache here?

To prevent redundant creation of multiple reference objects that reference the same variable/function in the kernel.

For example,
```python
@testbook('notebook.ipynb')
def foo(tb):
    assert tb.foo(1, 2) == 3 # new reference obj created
    assert tb.foo(2, 3) == 5 # yet another obj created
```

Reply:
You're not going to slow anything down by creating new reference objects here. I think we should ditch caching until there's an actual bottleneck. i.e. [premature-optimization root-of-evil](https://stackify.com/premature-optimization-evil/#:~:text=%E2%80%9CPremature%20optimization%20is%20the%20root,famous%20saying%20among%20software%20developers.&text=%E2%80%9CThe%20real%20problem%20is%20that,of%20it)%20in%20programming.%E2%80%9D)